src/jltypes.c

```diff
@@ -2075,30 +1909,139 @@ static void cache_insert_type(jl_value_t *type, ssize_t insert_at, int ordered)
     }
 }
 
-jl_value_t *jl_cache_type_(jl_datatype_t *type)
+jl_value_t *jl_cache_type_(jl_datatype_t *type)
+{
+    if (is_cacheable(type)) {
+        JL_TIMING(TYPE_CACHE_INSERT);
+        int ord = is_typekey_ordered(jl_svec_data(type->parameters), jl_svec_len(type->parameters));
+        ssize_t idx = lookup_type_idx(type->name, jl_svec_data(type->parameters),
+                                      jl_svec_len(type->parameters), ord);
+        if (idx >= 0)
+            type = (jl_datatype_t*)jl_svecref(ord ? type->name->cache : type->name->linearcache, idx);
+        else
+            cache_insert_type((jl_value_t*)type, ~idx, ord);
+    }
+    return (jl_value_t*)type;
+}
+
+// type instantiation
+
+static int valid_type_param(jl_value_t *v)
+{
+    if (jl_is_tuple(v)) {
+        // NOTE: tuples of symbols are not currently bits types, but have been
+        // allowed as type parameters. this is a bit ugly.
+        jl_value_t *tt = jl_typeof(v);
+        size_t i;
+        size_t l = jl_nparams(tt);
+        for(i=0; i < l; i++) {
+            jl_value_t *pi = jl_tparam(tt,i);
+            if (!(pi == (jl_value_t*)jl_sym_type || jl_isbits(pi)))
+                return 0;
+        }
+        return 1;
+    }
+    // TODO: maybe more things
+    return jl_is_type(v) || jl_is_typevar(v) || jl_is_symbol(v) || jl_isbits(jl_typeof(v));
+}
+
+static int within_typevar(jl_value_t *t, jl_tvar_t *v)
```

Comment:
From the name I initially assumed this meant some kind of containment. Add comment.